sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala

```diff
@@ -659,22 +659,46 @@ object CoGroup {
       right: LogicalPlan): LogicalPlan = {
     require(StructType.fromAttributes(leftGroup) == StructType.fromAttributes(rightGroup))
 
+    // SPARK-42132: The DeduplicateRelations rule would replace duplicate attributes
+    // in the right plan and rewrite rightGroup and rightAttr. But it would also rewrite
+    // leftGroup and leftAttr, which is wrong. Additionally, it does not rewrite rightDeserializer.
+    // Aliasing duplicate attributes in the right plan deduplicates them and stops
+    // DeduplicateRelations to do harm.
+    val duplicateAttributes = AttributeMap(
+      right.output.filter(left.output.contains).map(a => a -> Alias(a, a.name)())
+    )
+
+    def dedup(attrs: Seq[Attribute]): Seq[NamedExpression] =
+      attrs.map(attr => duplicateAttributes.getOrElse(attr, attr))
+
+    // rightOrder is resolved against right plan, so deduplication not needed
+    val (dedupRightGroup, dedupRightAttr, dedupRight) =
+      if (duplicateAttributes.nonEmpty) {
```

Comment:
it's a bit weird to do dedup here. Can we update the `DeduplicateRelations` rule to handle `CoGroup` specially?

Reply:
I don't really see how `DeduplicateRelations` can be modified to not rewrite **all** attributes of CoGroup.

In `DeduplicateRelations.apply` method `renewDuplicatedRelations` is called, which calls `rewriteAttrs(attrMap)` on the `CoGroup`, which rewrites **all** attributes.

If you are suggesting to add `case cogroup @ CoGroup(...) =>` to
https://github.com/apache/spark/blob/f1a6c5ed08a8f787b4924800f2a345b5a5c94dbf/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala#L45-L83
then this won't work because all attributes of `CoGroup` have been rewritten at this point.

Reply:
@Hisoka-X  is working on it: https://github.com/apache/spark/pull/41554